fiftyone/utils/utils3d.py

```diff
@@ -0,0 +1,204 @@
+"""
+3D utilities.
+
+| Copyright 2017-2022, Voxel51, Inc.
+| `voxel51.com <https://voxel51.com/>`_
+|
+"""
+import numpy as np
+import scipy.spatial as sps
+
+import eta.core.numutils as etan
+
+
+# Reference: https://github.com/AlienCat-K/3D-IoU-Python/blob/master/3D-IoU-Python.py
+def compute_cuboid_iou(gt, pred, gt_crowd=False):
+    """Computes the IoU between the given ground truth and predicted cuboids.
+
+    Args:
+        gt: a :class:`fiftyone.core.labels.Detection`
+        pred: a :class:`fiftyone.core.labels.Detection`
+        gt_crowd (False): whether the ground truth cuboid is a crowd
+
+    Returns:
+        the IoU, in ``[0, 1]``
+    """
+    corners1 = _3d_corners(gt)
+    corners2 = _3d_corners(pred)
+
+    # points are in counter-clockwise order
+    rect1 = [(corners1[i, 0], corners1[i, 2]) for i in range(3, -1, -1)]
+    rect2 = [(corners2[i, 0], corners2[i, 2]) for i in range(3, -1, -1)]
+
+    _, inter_area = _convex_hull_intersection(rect1, rect2)
+    ymax = min(corners1[0, 1], corners2[0, 1])
+    ymin = max(corners1[4, 1], corners2[4, 1])
+    inter_vol = inter_area * max(0.0, ymax - ymin)
+
+    # area1 = _polygon_area(np.array(rect1)[:, 0], np.array(rect1)[:, 1])
+    # area2 = _polygon_area(np.array(rect2)[:, 0], np.array(rect2)[:, 1])
+    # iou_2d = inter_area / (area1 + area2 - inter_area)
+
+    gt_vol = _cuboid_volume(corners1)
+    pred_vol = _cuboid_volume(corners2)
+
+    if gt_crowd:
+        union = pred_vol
+    else:
+        union = gt_vol + pred_vol - inter_vol
+
+    return min(etan.safe_divide(inter_vol, union), 1)
+
+
+def _3d_corners(detection):
+    rotation = np.array(detection.rotation)
+    location = np.array(detection.location)
+    x, y, z = np.array(detection.dimensions)
+    R = _rotation_matrix(rotation)
+
+    corners = np.vstack(
+        [
+            [x / 2, x / 2, -x / 2, -x / 2, x / 2, x / 2, -x / 2, -x / 2],
+            [z / 2, z / 2, z / 2, z / 2, -z / 2, -z / 2, -z / 2, -z / 2],
```

Comment:
I renamed `l, w, h` to `x, y, z` in this implementation, which makes it easier to see that you're swapping y and z coordinates here. Therefore, this method doesn't technically return 3D corners according to how we've defined the `dimensions`/`location`/`rotation`, right?

Since the unit tests are passing, I assume the rest of the implementation accounts for the swapped coordinates. But, I think it would be better to use `(x, y, z)` instead of `(x, z, y)`.